File: [N]greedy/[N]greedy/621-task-scheduler.py
from collections import defaultdict
class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        task_freq = defaultdict(int)
        for t in tasks:
            task_freq[t] += 1
        max_freq = max(task_freq.values())
        tasks_with_max_freq = []
        for t, f in task_freq.items():
            if f == max_freq:
                tasks_with_max_freq.append(t)
        
        h = max_freq - 1
        w = n + 1
        res = h * w + len(tasks_with_max_freq)
        return res if res > len(tasks) else len(tasks)

# time O(n)
# space O(1), due to task's category
# using greedy and hashmap

File: [N]greedy/[N]greedy/621-task-scheduler.java
class Solution {
    public int leastInterval(char[] tasks, int n) {
        HashMap<Character, Integer> charToFreq = new HashMap<>();
        for (int i = 0; i < tasks.length; i++) {
            charToFreq.compute(tasks[i], (k, v) -> v == null ? 1 : v + 1);
        }
        int maxFreq = Collections.max(charToFreq.values());
        int maxFreqCount = 0;
        for (Character c: charToFreq.keySet()) {
            if (Objects.equals(charToFreq.get(c), maxFreq)) {
                maxFreqCount += 1;
            }
        }
        return Math.max(tasks.length, (maxFreq - 1) * (n + 1) + maxFreqCount);
    }
}

// time O(n)
// space O(1), due to task's category
// using greedy and hashmap